contracts/Vault.sol

```diff
@@ -1,189 +1,101 @@
+pragma solidity ^0.8.0;
+import "./ERC4626Metadata.sol";
 import "@openzeppelin/contracts/token/ERC20/presets/ERC20PresetMinterPauser.sol";
-import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
-import "./IMarket.sol";
-import "./IVault.sol";
+import "@openzeppelin/contracts/interfaces/IERC20Metadata.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
 
-contract Vault is Ownable, IVault, ERC20PresetMinterPauser {
-    // ERC20
-    uint256 private _totalSupply;
-    mapping(address => uint256) private _balances;
+contract Vault is ERC4626Metadata, Ownable {
 
-    mapping(address => uint256) private _shares;
+    using Math for uint256;
 
-    IERC20Metadata private immutable _underlying;
-    address private immutable _self;
+    //Mapping address => uint256
+    mapping(address => uint256) public marketAllowance;
     
     // These will change to allow multiple markets
     address private _market;
     uint8 private immutable _decimals;
-
-    constructor(
-        IERC20Metadata underlying
-    )
-        ERC20PresetMinterPauser(
-            string(abi.encodePacked("HL ", underlying.name())),
-            string(abi.encodePacked("HL", underlying.symbol()))
-        )
-    {
+    address private _self;
+
+    constructor(IERC20Metadata asset_)
+    ERC4626Metadata(
+        asset_
+    ) 
+    ERC20(
+        string(abi.encodePacked("HL ", asset_.name())),
+        string(abi.encodePacked("HL", asset_.symbol()))
+    ) {
         require(
-            address(underlying) != address(0),
+            address(asset_) != address(0),
             "Underlying address is invalid"
         );
-
+        _decimals = IERC20Metadata(asset_).decimals();
         _self = address(this);
-        _underlying = underlying;
-        _decimals = IERC20Metadata(underlying).decimals();
     }
 
     // Override decimals to be the same as the underlying asset
-    function decimals() public view override(ERC20) returns (uint8) {
+    function decimals() public view override (ERC20) returns (uint8) {
         return _decimals;
     }
 
-    function totalSupply()
-        public
-        view
-        override
-        returns (uint256)
-    {
-        return _totalSupply;
-    }
-
-    function getMarketAllowance() external view returns (uint256) {
-        // // TODO: This will change to allow multiple markets, using msg.sender
-        uint256 allowance = _underlying.allowance(_self, _market);
-        if (allowance > _totalAssets()) {
-            return _totalAssets();
-        }
-        
-        return allowance;
-    }
-
-    function getMarket() external view returns (address) {
-        return _market;
-    }
-
     function setMarket(address market, uint256 max) public onlyOwner {
         require(_market == address(0), "setMarket: Market already set");
-
         _market = market;
+        IERC20(asset()).approve(_market, max);
```

Comment:
do we need to pass in `max`? im assuming this is the max `unit256` size? or are we intending to pass in any max approval amount?

Reply:
this is just the way it currently works on the main branch. It will be changing soon.